### PR TITLE
Fix flaky test `TestGetContainerStatsRmRunning`

### LINKED ISSUE #18679: Flaky test: TestGetContainerStatsRmRunning
From #18617 on Windows:

```
----------------------------------------------------------------------
FAIL: docker_api_containers_test.go:287: DockerSuite.TestGetContainerStatsRmRunning

docker_api_containers_test.go:321:
    c.Assert(err, checker.Not(checker.IsNil))
... value = nil


----------------------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by WeiZhang555**
Fixes #18679

It's hard to reproduce the error, so I'm not 100% sure that this is a perfect fix.
Hope it will really help.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

### CODE REVIEW

**Comment by WeiZhang555** [comment]
Just found a way to reproduce the flaky error, only need to add one line of code before https://github.com/docker/docker/pull/19570/files#diff-ca97ecb464e792b1086fca2ee8af23f8R319
to make it looks like:

```
 	c.Assert(err, checker.IsNil)
        <-time.After(2 * time.Second)     // add this line
  	dockerCmd(c, "rm", "-f", id)
```
Two seconds' suspend is enough for reproducing the error, this will help prove the fix.
So I'm 99% sure about the correctness of this fix now :)

**Comment by coolljt0725** [comment]
This test case is introduced in #13499,  ping @cpuguy83

File: integration-cli/docker_api_containers_test.go [hunk 1]
```diff
@@ -318,6 +318,19 @@ func (s *DockerSuite) TestGetContainerStatsRmRunning(c *check.C) {
 	c.Assert(err, checker.IsNil)
 	dockerCmd(c, "rm", "-f", id)
 
+	// Flush cache and make sure there's really nothing in the channel
+	var stop bool
+	flushChannel := func() {
+		var errFlush error
+		for errFlush == nil && !stop {
```
Comment (cpuguy83) [inline integration-cli/docker_api_containers_test.go]: this is pretty racey

**Comment by cpuguy83** [comment]
I'm thinking the fix for this is a little different.

At https://github.com/docker/docker/blob/master/integration-cli/docker_api_containers_test.go#L293, let's move that chan out into it's own variable.
Remove defer from here: https://github.com/WeiZhang555/docker/blob/18679-flaky-test/integration-cli/docker_api_containers_test.go#L294

Then once `io.Copy` is finished, close the buf.
At the end of the test we just make sure the channel is closed: `_, open := <-bufChan; open { c.Fatal("expected buffer to be closed") }`

Really the last bit is just a sanity check.

**Comment by cpuguy83** [comment]
WDYT?

**Comment by WeiZhang555** [comment]
@cpuguy83 Already made some updates according to your ideas, can you take a look to see if this is what you are thinking? 

I still keep the flushing channel part because I am really not sure how to handle the cache.

File: pkg/integration/utils.go [hunk 2]
```diff
@@ -344,6 +344,21 @@ func (c *ChannelBuffer) ReadTimeout(p []byte, n time.Duration) (int, error) {
...
 
+// FlushTimeout flush content of channel within the specified duration as timeout
+func (c *ChannelBuffer) FlushTimeout(n time.Duration) error {
+	var stop bool
+	go func() {
+		<-time.After(n)
+		stop = true
+	}()
+	for range c.C {
+		if stop {
```
Comment (cpuguy83) [inline pkg/integration/utils.go]: Reading and writing to the same variable without synchronization will be racey and could cause a panic.
Also not sure what this is flush is doing, I don't see where we are flushing.

**Comment by cpuguy83** [comment]
I don't think we need to flush it, once it gets to that point we don't care about any of the data in the buffer anymore.

**Comment by WeiZhang555** [comment]
@cpuguy83 
So I want to share some my thoughts about why this test case turn out to be flaky, I have mentioned the reproduce steps:
```
 c.Assert(err, checker.IsNil)
    <-time.After(2 * time.Second)     // add this line
    dockerCmd(c, "rm", "-f", id)
```

The reason is before `docker rm -f id`, `docker stats` will probably get a chance to print another stats result into channel, so even after the container is forced to be removed, the channel will still keep something even it is closed.

So the original judgement can't work:
```
-	_, err = buf.ReadTimeout(b, 2*time.Second)
 -	c.Assert(err, checker.Not(checker.IsNil))
```
because it will first read cache content in channel(the channel has a buffer length of 1)
Of course it often works, because `docker stats` didn't get chance to print something into channel before `docker rm -f`
Also 
```
if _, open := <-bufChan; open { 
     c.Fatal("expected buffer to be closed") 
}
```
can't work, because even it's closed, `open` will still be true if channel has content. 
We can change it to 
```
<-bufChan    //this is for flushing cache
if _, open := <-bufChan; open { 
     c.Fatal("expected buffer to be closed") 
}
```
This may work.
But I am not sure with the channel behavior, so what I was trying to do is read and discard all contents from channel in spite of how many caches are there, after the channel is empty, we can 
judge whether it's closed or not safely.

File: pkg/integration/utils.go (hunk 2 above)
Comment (WeiZhang555) [inline pkg/integration/utils.go]: I think the problem may not exists. 
We also don't have synchronization mechanism between `Write` and `ReadTimeout`, it also works fine.
I have to say I'm a little confused about the ChannelBuffer and `C chan[]byte`, I'm not very confident about its behavior. 
I think that's also why you and I have different understanding about it.

**Comment by WeiZhang555** [comment]
Quick update:
```
<-bufChan    //this is for flushing cache
if _, open := <-bufChan; open { 
     c.Fatal("expected buffer to be closed") 
}
```
Also doesn't work, still be flaky according my test.

I think we still need the flush function.

**Comment by cpuguy83** [comment]
@WeiZhang555 You're right.
Maybe we should just get rid of everything after line 322 since the test was really testing that stats are not removed from a failed `docker rm`.

I don't think we can verify the rest of it without being racey.

**Comment by WeiZhang555** [comment]
@cpuguy83 Yes, I agree. It's racey and hard to verify and avoid. If the test cases was only intending to check against a failed `docker rm`, then I think it's OK to remove the rest part.

It's all up to you, I'll do that if you are also :+1: on removing code after 322. :wink:

**Comment by cpuguy83** [comment]
Yep, SGTM.
Thanks for working on this!

**Comment by WeiZhang555** [comment]
@cpuguy83 Done, please take a look at whether I make it right.

**Comment by WeiZhang555** [comment]
ping @cpuguy83

**Comment by cpuguy83** [comment]
LGTM

**Comment by icecrime** [comment]
LGTM

**Pull request merged**